### PR TITLE
ARROW-12722: [R] Raise error when attemping to print table with duplicated naming

### PR DESCRIPTION
**Pull request by thisisnic**

### CODE REVIEW

File: r/R/dplyr.R [hunk 1]
```diff
@@ -74,7 +74,13 @@ print.arrow_dplyr_query <- function(x, ...) {
     name <- expr$field_name
     if (nzchar(name)) {
       # Just a field_ref, so look up in the schema
-      schm$GetFieldByName(name)$type$ToString()
+      field_name <- schm$GetFieldByName(name)
```
Comment (nealrichardson) [inline r/R/dplyr.R:77]: I checked with @bkietz and (currently at least) evaluating expressions on a dataset with duplicated field names will error, so if that's the case (perhaps you could verify), we should check this sooner than the print method. How about instead we put a check around L33 above, like

```r
if (anyDuplicated(names(.data))) {
  abort("useful message")
}
```

**Review by nealrichardson — CHANGES_REQUESTED** [review]
I think this should fail earlier than the print method

File: r/R/dplyr.R (hunk 1 above)
Comment (thisisnic) [reply r/R/dplyr.R]: Have updated now

File: r/R/dplyr.R [hunk 2]
```diff
@@ -31,6 +31,18 @@ arrow_dplyr_query <- function(.data) {
...
+  if (anyDuplicated(names(.data))) {
+    abort(c(
+      "Duplicated field names",
+      x = paste0(
+        "The following field names were found more than once in the data: ",
+        paste(names(.data)[duplicated(names(.data))], collapse = ", ")
+      ),
+      i =  "All field names must be unique"
+    ))
+  }
```
Comment (nealrichardson) [inline r/R/dplyr.R:44]: Not that it matters much but we can call names() and duplicated() less (and use our paste helper):

```suggestion
  dupes <- duplicated(names(.data))
  if (any(dupes)) {
    abort(c(
      "Duplicated field names",
      x = paste0(
        "The following field names were found more than once in the data: ",
        oxford_paste(names(.data)[dupes])
      ),
      i =  "All field names must be unique"
    ))
  }
```

Also, do we need `i =  "All field names must be unique"`? Is that something that `abort()` suggests/insists on? It seems redundant.

**Review by nealrichardson — APPROVED** [review]
A question and a suggestion for you but otherwise LGTM, thanks!

File: r/R/dplyr.R (hunk 1 above)
Comment (bkietz) [reply r/R/dplyr.R]: Specifically: looking up fields by name when that name is not unique in the schema is not supported.

File: r/R/dplyr.R (hunk 2 above)
Comment (thisisnic) [reply r/R/dplyr.R]: Fair comment, it does not, and you're right it's redundant - have removed it now.

File: r/tests/testthat/test-dplyr.R [hunk 3]
```diff
@@ -593,3 +594,12 @@ test_that("bad explicit type conversions with as.*()", {
   )
 
 })
+
+test_that("No duplicate field names are allowed in an arrow_dplyr_query", {
+  expect_error(
+    Table$create(tbl, tbl) %>%
+      filter(int > 0),
+    regexp = "The following field names were found more than once in the data: int, dbl, dbl2, lgl, false, chr, fct, verses, padded_strings"
```
Comment (nealrichardson) [inline r/tests/testthat/test-dplyr.R:602]: The last change altered the error message slightly so you'll need to fix (I'd make a `suggestion` here but you'll still need to pull and run to confirm)

**Pull request closed without merging**